src/passes/OptimizeInstructions.cpp

```diff
@@ -3397,6 +3401,67 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  Expression* simplifyRoundingsAndConversions(Unary* unaryOuter) {
+    switch (unaryOuter->op) {
+      case TruncSFloat64ToInt32:
+      case TruncSatSFloat64ToInt32: {
+        // i32 -> f64 -> i32 rountripping optimization:
+        //   i32.trunc(_sat)_f64_s(f64.convert_i32_s(x))  ==>  x
+        if (auto* unaryInner = unaryOuter->value->dynCast<Unary>()) {
```

Comment:
Can we use `matches` for at least the inner expressions here?

Reply:
For inner cases, I guess it makes sense.

Reply:
@tlively That's quite strange but it seems `matches` macros does not work inside `case`. I rewrote all inner rules using matches and all the rules just didn't work out.

Reply:
```c++
   switch (curr->op) {
      case TruncSFloat64ToInt32:
      case TruncSatSFloat64ToInt32: {
        // i32 -> f64 -> i32 rountripping optimization:
        //   i32.trunc(_sat)_f64_s(f64.convert_i32_s(x))  ==>  x
        Expression* x;
        if (matches(curr, unary(ConvertSInt32ToFloat64, any(&x)))) {
          return x;
        }
        break;
      }
      case TruncUFloat64ToInt32:
      case TruncSatUFloat64ToInt32: {
        // u32 -> f64 -> u32 rountripping optimization:
        //   i32.trunc(_sat)_f64_u(f64.convert_i32_u(x))  ==>  x
        Expression* x;
        if (matches(curr, unary(ConvertUInt32ToFloat64, any(&x)))) {
          return x;
        }
        break;
      }
```

Reply:
`matches` is a function, not a macro, so it shouldn't cause any problems. You do need the braces `{}` in the cases to declare new local variables, though. I think the problem above is that you should be matching `curr->value` rather than `curr`.

Reply:
Ah, you're right! Done!